### PR TITLE
[DMA 6/8] More helper types & working `split`

### LINKED ISSUE #2401: Cannot work with DMA channels in a generic way
## Bug description

I'm trying _not_ to hard-code peripherals and work with their all-generic versions, yet I fail when trying to swap-in a generic channel type for where a DMA channel was passed previously.

Consider this snippet:
```rust
use esp_hal::dma;
use esp_hal::dma::*;
use esp_hal::dma_buffers;
use esp_hal::gpio;
use esp_hal::peripheral::Peripheral;
use esp_hal::peripherals;
use esp_hal::prelude::*;
use esp_hal::spi;
use esp_hal::Async;

fn new_foo<'d, T, const N: u8>(
    dma_channel: dma::ChannelCreator<N>, // PROBLEM: I want this to be generic
    spi: impl Peripheral<P = T> + 'd,
    sclk: impl Peripheral<P = impl gpio::OutputPin> + 'd,
    sdo: impl Peripheral<P = impl gpio::OutputPin> + 'd,
    sdi: impl Peripheral<P = impl gpio::InputPin> + 'd,
    cs: impl Peripheral<P = impl gpio::OutputPin> + 'd,
    reset: impl Peripheral<P = impl gpio::OutputPin> + 'd,
    ready: impl Peripheral<P = impl gpio::InputPin> + 'd,
)
where 
    T: spi::master::InstanceDma,
    dma::ChannelCreator<N>: esp_hal::dma::DmaChannelConvert<<T as esp_hal::dma::DmaEligible>::Dma>,
{
    let dma_channel = dma_channel.configure_for_async(false, DmaPriority::Priority0);

    let (rx_buffer, rx_descriptors, tx_buffer, tx_descriptors) = dma_buffers!(32000);
    let dma_rx_buf = dma::DmaRxBuf::new(rx_descriptors, rx_buffer).unwrap();
    let dma_tx_buf = dma::DmaTxBuf::new(tx_descriptors, tx_buffer).unwrap();
            
    let spi_driver = spi::master::Spi::new(spi, 10.MHz(), spi::SpiMode::Mode0)
        .with_sck(sclk)
        .with_mosi(sdo)
        .with_miso(sdi)
        .with_dma(dma_channel)
        .with_buffers(dma_rx_buf, dma_tx_buf);
}
```

Multiple issues with it:
1. `dma_channel` is moved into the function, whereas I want to keep the flexibility that it is either moved or mutably borrowed (and can be reused when the `spi_driver` instance gets dropped)
2. Even this current ^^^ syntax does not work, because `configure_for_async` is defined (with a macro) for each concrete `ChannelCreator<1, 2, 3, ...4>` rather than for `impl<const N: u8> ChannelCreator<N>`
3. The above code needs this weird `dma::ChannelCreator<N>: esp_hal::dma::DmaChannelConvert<<T as esp_hal::dma::DmaEligible>::Dma>` to work. That is, given that the issue (2.) is solved

## To Reproduce

Try to type-check the above code.

## Expected behavior

Ideally, I should be able to type something like:
`dma_channel: impl Peripheral<P = impl SomethingSomethingChannelCreatorSomething> + 'd` and be done with it.

## Environment

 I'm working with esp32s3, but I think the problem is generic. Latest released `esp-hal` as of today (0.21.1) but the issue persist with `main`.

### PR DESCRIPTION
**Pull request by bugadani**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description

Closes #2401
Closes https://github.com/esp-rs/esp-hal/issues/1767

### CODE REVIEW

**Comment by bugadani** [comment]
I'm open to suggestions for the naming. I'm aware that type erasure hurts us a bit on 32 and S2. I am planning to use different erased types on the P4 so we won't have runtime checks there.

**Comment by Dominaezzz** [comment]
This also closes https://github.com/esp-rs/esp-hal/issues/1767

**Comment by bugadani** [comment]
I'm a bit unsure whether this amount of documentation is too little, enough or too much, but I guess it's time to collect feedback.

File: esp-hal/src/dma/mod.rs [hunk 1]
```diff
@@ -1657,6 +1657,91 @@ impl<DEG: DmaChannel> DmaChannelConvert<DEG> for DEG {
...
+#[doc = ""]
+/// let spi = Spi::new_with_config(
+///     peripherals.SPI2,
+///     Config::default(),
+/// );
+///
+/// let spi_dma = configures_spi_dma(spi, dma_channel);
+/// # }
+/// ```
+pub trait CompatibleWith<P: DmaEligible>: DmaChannel + DmaChannelConvert<DmaChannelFor<P>> {}
```
Comment (Dominaezzz) [inline esp-hal/src/dma/mod.rs:1707]: Just to throw my vote in name-wise. I think `DmaChannelFor` is a better name than `CompatibleWith` for what users will be writing.

I haven't look into the consequences of making this happen :smile:, was just my two cents on the public name.

2 cents aside.

Are these still needed now? Seems like users won't be using it directly anymore.
```Rust
/// Helper type to get the DMA (Rx and Tx) channel for a peripheral.
pub type DmaChannelFor<T> = <T as DmaEligible>::Dma;
/// Helper type to get the DMA Rx channel for a peripheral.
pub type RxChannelFor<T> = <DmaChannelFor<T> as DmaChannel>::Rx;
/// Helper type to get the DMA Tx channel for a peripheral.
pub type TxChannelFor<T> = <DmaChannelFor<T> as DmaChannel>::Tx;
```

**Review by Dominaezzz — COMMENTED** [review]
Haven't read the doc yet, just one comment

File: esp-hal/src/dma/mod.rs (hunk 1 above)
Comment (bugadani) [reply esp-hal/src/dma/mod.rs:1707]: > Seems like users won't be using it directly anymore.

Not the users, but we are, and they really help :D But this naming clashes with the `CompatibleWith` types but they do different things 🤔 I don't really know what would be best.

Comment (Dominaezzz) [reply esp-hal/src/dma/mod.rs:1707]: > Not the users, but we are, and they really help

Ah true, #[docs(hidden)] then? Rust might actually allow them to be `pub(crate)` but not sure.

I'll leave everyone else to pitch in on the names.

Comment (bugadani) [reply esp-hal/src/dma/mod.rs:1707]: I think I can merge them into Channel so that we'd have `Channel<'d, DmaMode, Peripheral>`

**Comment by bugadani** [comment]
> I'll leave it you to decide when to add split() to the migration guide.

Thanks, forgot to add it

File: esp-hal/src/dma/mod.rs (hunk 1 above)
Comment (bugadani) [reply esp-hal/src/dma/mod.rs]: Certainly a lot better than `ChannelCreator<2>: DmaChannelConvert<<T as DmaEligible>::Dma>,` 🙃

**Pull request merged**